design/user-event-spi.md

```diff
@@ -0,0 +1,73 @@
+# User Event SPI
+
+* **Status**: Draft v0.1
+* **Author**: [vmuzikar](https://github.com/vmuzikar)
+
+## Motivation
+
+Current Event SPI design has several limitations. E.g. it's not defined when an event is fired exactly, there are no pre-
+and post-events. There's also no way of pre-events intercepting for approve/reject functionality.
+
+## Use cases
+
+* A user is removed from Keycloak. All related records in an external system are removed before the user is permanently
+  deleted (e.g. eshop orders).
+* A user self-registers in Keycloak. Their registration must be approved by an external system (either human or automated)
+  before the new account is actually fully created.
+* A user self-edits their email address. The change must be approved first by an external system before it's actually
+  applied.
+
+## Design proposal
+
+To overcome the current Event SPI limitations, a new "domain-specific SPI" will be implemented – User Event SPI.
+
+Key changes over current generic Event SPI:
+* Each event is fired twice – as a pre- and post-event.
+* Pre-events are "interactive" – listener can approve, reject or delegate (for later decision by an external system)
+  each pre-event. 
+* The same User Event SPI will be used everywhere – admin console and REST endpoints, user registration form,
+  account console and REST endpoints, etc.
+* The event types will be much more detailed, not just usual CRUD. There'll be separated events for user self-registration,
+  creation through Admin REST API, Account Console, etc.
+
```

Comment:
I would love to see some prototype written as code here. How the interface would look like. What would be the events definition etc.

Reply:
Sure, I plan to add some list of the new events as well as some code snippets (probably on Friday).

Reply:
@slaskawi I've added the list of event types as well as some code snippets. Those are really high-level and from the perspective of a Keycloak dev.

Reply:
Thanks!